### ISSUE TITLE
win_regedit: change detection doesn't work for environment variables in registry keys of type expandstring

### ISSUE THREAD

**Issue by chrwen-omicron** [ansible/ansible#19213]
_From @chrwen-omicron on June 17, 2016 14:2_

##### ISSUE TYPE

- Bug Report
##### COMPONENT NAME

win_regedit
##### ANSIBLE VERSION

```
ansible 2.1.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
```
##### SUMMARY

Setting an environment variable as data of an expandstring data type leads to a change every run.
##### STEPS TO REPRODUCE

```
- name: set an environment variable as data
  win_regedit:
    key: "HKCU:\\Environment"
    value: HOME
    data: "%USERPROFILE%"
    type: expandstring
```

##### EXPECTED RESULTS

No change if the registry contains `%USERPROFILE%.`
##### ACTUAL RESULTS

When win_regedit is reading the registry key, it gets the expanded data instead of the real data.
E.g. `C:\Users\<username>` instead of `%USERPROFILE%`

_Copied from original issue: ansible/ansible-modules-extras#2446_

**Comment by chrwen-omicron** [comment]
_From @jhawkesworth on June 17, 2016 17:2_

A change was merged into devel yesterday that affected change detection for win_regedit.
Please could you try latest devel?

**Comment by chrwen-omicron** [comment]
Sorry for not mentioning this, I did try the most recent version of win_regedit in the devel branch.

The problem seems to be `Get-ItemProperty` in

``` powershell
$currentRegistryData =(Get-ItemProperty -Path $registryKey | Select-Object -ExpandProperty $registryValue)
```

This returns the expanded string.

**Comment by chrwen-omicron** [comment]
_From @jhawkesworth on June 20, 2016 11:2_

thanks @chrwen-omicron - Thinking about it further, I don't have any occurrences of expand string in my playbooks right now, and I'm pretty certain this isn't covered in the existing integration tests.

I suspect the type of $currentRegistryData is not one that is being handled in Compare-RegistryData

Could you find the type of $currentRegistryData when working with an expandstring?

```
write-host $currentRegistryData.GetType().Name 
```

would get it.  Might be easiest to set ANSIBLE_KEEP_REMOTE_FILES=1 on your controller, run the playbook once then re-run in Powershell ISE on a windows node otherwise write-host likely would get lost.

If it isn't a String, int or Object[] then the compare test won't handle it at the moment.

```
        if ($ReferenceData -is [String] -or $ReferenceData -is [int]) {
            if ($ReferenceData -eq $DifferenceData) {
                return $true
            } else {
                return $false
            }
        } elseif ($ReferenceData -is [Object[]]) {
            if (@(Compare-Object $ReferenceData $DifferenceData -SyncWindow 0).Length -eq 0) {
                return $true
            } else {
                return $false
            }
        }
```

If it is one of the existing types then its possible we are not comparing the right thing - because what we are storing in the ansible configuration isn't the final value, just something that can be expanded out to the right value, we might not be able to make a valid comparison to see whether it has changed.

Also, win_environment exists if you want to manage environment variables.

Hope this helps,

Jon

**Comment by chrwen-omicron** [comment]
Well, the registry types REG_SZ and REG_EXPAND_SZ seem to be basically the same. They're both strings, so the compare test is fine, but REG_EXPAND_SZ may contain unexpanded references to environment variables.
`if ($ReferenceData -is [String] -or $ReferenceData -is [int])` is handling it correctly, but the `$ReferenceData` is expanded and the `$DifferenceData` not, so they differ at `if ($ReferenceData -eq $DifferenceData)`.

I tried to use win_environment in the first place, but ended up using win_regedit because it doesn't support expandstring.

This line seems to do the trick and returns `%USERPROFILE%` instead of `C:\Users\<username>`:

``` powershell
(Get-Item -Path $registryKey).GetValue($registryValue, "", "DoNotExpandEnvironmentNames")
```

But I don't know powershell, so this might have side effects which I'm not aware of.

Regards

**Comment by chrwen-omicron** [comment]
_From @ansibot on August 4, 2016 15:52_

@joshludwig, @smadam813, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by chrwen-omicron** [comment]
_From @ansibot on September 11, 2016 15:1_

@joshludwig, @smadam813, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by chrwen-omicron** [comment]
_From @jhawkesworth on September 19, 2016 16:48_

I think I am against making a change to support this as it would mean that you couldn't look at your ansible config and determine ahead of time for certain what the final registry value would be. I can see how it is desirable from the point of view of being automatable, but it stops ansible configuration being a 'single source of truth', which I value. I get that it can make it harder to accommodate the many and varied needs of your system users.

@chrwen-omicron - 
Perhaps you could give an example use case to make it easier to reason about?

**Comment by chrwen-omicron** [comment]
@jhawkesworth following your argument you're not allowed to use environment variables at all in your ansible config.

The important thing here is that the write operation behaves differently from the read operation. The write can handle unexpanded environment variables for the REG_EXPAND_SZ type whereas the read can't.
At least you should get an option to select the behavior.

This is also a problem if you want to extend your PATH variable for instance, because you're expanding the variables inside the PATH.

**Comment by chrwen-omicron** [comment]
_From @ansibot on October 6, 2016 17:56_

@joshludwig, @smadam813, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by chrwen-omicron** [comment]
_From @ansibot on October 22, 2016 10:14_

@joshludwig, @smadam813, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by chrwen-omicron** [comment]
_From @ansibot on November 6, 2016 10:49_

@joshludwig, @smadam813, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by chrwen-omicron** [comment]
_From @ansibot on November 22, 2016 10:57_

@joshludwig, @smadam813, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by chrwen-omicron** [comment]
_From @ansibot on December 6, 2016 22:1_

This repository has been locked. All new issues and pullrequests should be filed in https://github.com/ansible/ansible

Please read through the <a href="http://docs.ansible.com/ansible/dev_guide/repomerge.html#move-issues-and-prs-to-new-repo">repomerge</a> page in the dev guide. The guide contains links to tools which automatically move your issue or pullrequest to the ansible/ansible repo.

**Comment by dagwieers** [comment]
@chrwen-omicron The win_regedit module has been refactored quite a bit.
Can you please test if your issue still stands using the devel branch ?

**Comment by chrwen-omicron** [comment]
@dagwieers It won't work as long as `$old_data = (Get-ItemProperty -Path $path | Select-Object -ExpandProperty $name)` returns the expanded data.

**Comment by dagwieers** [comment]
@chrwen-omicron Please propose a PR with the changes you require. Or provide me with a set of integration tests that indicate the problem to me. Because I think I know what you are saying, but I want to be clear that if I am going to fix this, I am covering all grounds.

What option would you suggest to change the behaviour (i.e. expand / do not expand data) ?
What is your preferred way of doing this ? And how would it behave by default.

**Comment by jhawkesworth** [comment]
needs_info

**Comment by ansibot** [comment]
@chrwen-omicron This issue is waiting for your response. Please respond or the issue will be closed.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by chrwen-omicron** [comment]
@dagwieers I created #22874 to reproduce the problem.

**Comment by dagwieers** [comment]
At [19:00 UTC](http://www.thetimezoneconverter.com/?t=19:00&tz=UTC) this ticket will be processed during [the first Windows Sprint](https://github.com/ansible/community/wiki/Windows:-sprints#2017-07-04-1900-utc-sprint-1), we would appreciate if you could join the Sprint to help out with fixing/closing this issue.

**Comment by chrwen-omicron** [comment]
@dagwieers I'm afraid I can't participate because I've an appointment. But I think the pull request #22874 shows the problem pretty well.

**Comment by dagwieers** [comment]
@chrwen-omicron Sure, this was a boiler-plate message, we will look into the details of each issue later today. Thanks for your help !

**Comment by jborean93** [comment]
Hey @chrwen-omicron I've raised a PR to fix this issue here https://github.com/ansible/ansible/pull/26468. If you get some time please give it a test and let us know on the PR page if it works for you. I've manually tested it with that PR and can confirm it doesn't report a change on the 2nd time round.

Test Playbook
```
---
- name: test win_regedit stuff
  hosts: 2012R2
  tasks:
  - name: create new entry
    win_regedit:
      path: HKCU:\SOFTWARE\Test
      name: HOME
      data: "%USERPROFILE%"
      type: expandstring
      state: present

  - name: test idempotency
    win_regedit:
      path: HKCU:\SOFTWARE\Test
      name: HOME
      data: "%USERPROFILE%"
      type: expandstring
      state: present
```

Test Output
```
(ansible-py27) (ansible-py27) [jborean@jordan-fedora module-tester]$ ansible-playbook -i inventory.ini win_regedit_tests.yml

PLAY [test win_regedit stuff] ********************************************************************************************************************************

TASK [Gathering Facts] ***************************************************************************************************************************************
Thursday 06 July 2017  11:25:37 +1000 (0:00:00.025)       0:00:00.025 *********
ok: [Server2012R2]

TASK [create new entry] **************************************************************************************************************************************
Thursday 06 July 2017  11:25:39 +1000 (0:00:02.282)       0:00:02.307 *********
changed: [Server2012R2]

TASK [test idempotency] **************************************************************************************************************************************
Thursday 06 July 2017  11:25:40 +1000 (0:00:00.700)       0:00:03.008 *********
ok: [Server2012R2]

PLAY RECAP ***************************************************************************************************************************************************
Server2012R2               : ok=3    changed=1    unreachable=0    failed=0
```

**Comment by dagwieers** [comment]
resolved_by_pr #26468
